### PR TITLE
Auth: Rotate token patch

### LINKED ISSUE #56994: Grafana Login Timeout (v9.2.0)
**What happened**:
Upgraded to v9.2.0

**What you expected to happen**:
No change in login / timeout behavior.

**How to reproduce it (as minimally and precisely as possible)**:
After upgrading to v9.2.0 Grafana will log out of the active user account in approximately 60 seconds.

**Anything else we need to know?**:
I did review the Grafana setting and preferences and did not see any option to increase the timeout length, or to set it to infinite. 

**Environment**:
- Grafana version: 9.2.0
- Data source type & version: InfluxDB
- OS Grafana is installed on: Raspberry Pi 10 (32-bit)
- User OS & Browser: Mac OSX / Chrome
- Grafana plugins: Telegraf
- Others: NA

### PR DESCRIPTION
**Pull request by mgyongyosi**
**What is this feature?**
This PR fixes the rotate token functionality by using a `singleflight.Group` and making sure that only 1 thread updates the token when multiple parallel request are trying to rotate the same token at the same time. It meant to be an improvement for the short term and we are planning to refactor the whole session token handling part.

**Why do we need this feature?**

**Which issue(s) does this PR fix?**:

Fixes #56994

**Special notes for your reviewer**:

### CODE REVIEW

**Review by Jguer — APPROVED** [review]

File: pkg/services/auth/authimpl/auth_token.go [hunk 1]
```diff
@@ -268,83 +275,102 @@ func (s *UserAuthTokenService) lookupToken(ctx context.Context, unhashedToken st
...
-		needsRotation = rotatedAt.Before(now.Add(-urgentRotateTime))
+	type rotationResult struct {
+		rotated  bool
+		newToken *auth.UserToken
 	}
 
-	if !needsRotation {
-		return false, nil
-	}
+	rotResult, err, _ := s.singleflight.Do(fmt.Sprintf("%d-%s", model.Id, model.AuthToken), func() (interface{}, error) {
```
Comment (mgyongyosi) [inline pkg/services/auth/authimpl/auth_token.go:295]: Consider using only the `model.Id`.

File: pkg/services/contexthandler/contexthandler.go [hunk 2]
```diff
@@ -569,14 +573,15 @@ func (h *ContextHandler) rotateEndOfRequestFunc(reqContext *contextmodel.ReqCont
 		}
 
 		// FIXME (jguer): rotation should return a new token instead of modifying the existing one.
```
Comment (mgyongyosi) [inline pkg/services/contexthandler/contexthandler.go:575]: Remove this comment.

File: pkg/services/auth/authimpl/auth_token.go (hunk 1 above)
Comment (mgyongyosi) [reply pkg/services/auth/authimpl/auth_token.go]: Done

File: pkg/services/contexthandler/contexthandler.go (hunk 2 above)
Comment (mgyongyosi) [reply pkg/services/contexthandler/contexthandler.go]: Done

**Review by kalleep — APPROVED** [review]

**Pull request merged**